File: 17-LetterCombinationsOfaPhoneNumber.cs
// 17. Letter Combinations of a Phone Number

public class Solution {
    private IList<string> oList = new List<string>();
    private Dictionary<char, List<char>> map = new Dictionary<char, List<char>>();
    
    public IList<string> LetterCombinations(string digits) {
        if (digits == null || digits == string.Empty)
            return oList;
        
        map.Add('2', new List<char>() { 'a', 'b', 'c' });
        map.Add('3', new List<char>() { 'd', 'e', 'f' });
        map.Add('4', new List<char>() { 'g', 'h', 'i' });
        map.Add('5', new List<char>() { 'j', 'k', 'l' });
        map.Add('6', new List<char>() { 'm', 'n', 'o' });
        map.Add('7', new List<char>() { 'p', 'q', 'r', 's' });
        map.Add('8', new List<char>() { 't', 'u', 'v' });
        map.Add('9', new List<char>() { 'w', 'x', 'y', 'z' });
        
        builder(digits, 0, new StringBuilder());
        
        return oList;
    
    }
    
    private void builder(string digits, int i, StringBuilder cur)
    {
        
        if (i == digits.Length)
            oList.Add(cur.ToString());
        else
            foreach (var item in map[digits[i]])
            {
                cur.Append(item);
                
                builder(digits, i + 1, cur);
                
                cur.Remove(cur.Length - 1, 1);
            }
    }
}

File: LeetCode Java/14 Patterns/Backtracking/17-LetterCombinationsOfAPhoneNumber.java
/*
 * 17. Letter Combinations of a Phone Number
 * Given a string containing digits from 2-9 inclusive, return all possible letter combinations that the number could represent. Return the answer in any order.
 * A mapping of digit to letters (just like on the telephone buttons) is given below. Note that 1 does not map to any letters.
 * Example 1:
    * Input: digits = "23"
    * Output: ["ad","ae","af","bd","be","bf","cd","ce","cf"]
 */
class Solution {
    public List<String> letterCombinations(String digits) {
        LinkedList<String> output = new LinkedList(); // acting as a queue
        if (digits.length() == 0)
            return output;
        output.add("");

        String[] char_map = new String[] { "0", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };

        for (int i = 0; i < digits.length(); i++) {
            int index = Character.getNumericValue(digits.charAt(i));

            while (output.peek().length() == i) { // first element length vs i
                String permutation = output.remove(); // pop

                for (char c : char_map[index].toCharArray()) { // go through each letter
                    output.add(permutation + c); // re add to queue
                }
            }

        }

        return output;
    }
}